### DISCUSSION TITLE
RefValidator doesn't work

### DISCUSSION THREAD

**Discussion by vitno1** [networknt/json-schema-validator#833] [Q&A]
Hello
I have this schema with $ref
```
{
  "$schema": "https://json-schema.org/draft-07/schema#",
  "title": "1234 some title",
  "type": "object",
  "properties": {
    "rowFirst": {
      "$ref": ".1234.json#/row"
    },
    "rowSecond": {
      "$ref": ".1234.json#/row"
    }
  },
  "required": [
    "rowFirst",
    "rowSecond"
  ]
}

```
And also I have a schema which I refer to
```
{
  "$schema": "https://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "row": {
      "type": "array",
      "title": "1234 someTitle",
      "items": {
        "properties": {
          "id": {
            "title": "identifier",
            "description": "some description",
            "type": "integer"
          },
          "year": {
            "title": "born year",
            "description": "",
            "type": "integer"
          }
        },
        "required": [
          "id",
          "year"
        ],
        "maxProperties": 2
      }
    }
  }
}

```

I am trying to validate the following input data

```
{
  "rowFirst": [
    {
      "year": 2023
	},
    {
      "year": 2023
	}
  ],
  "rowSecond": [
    { 
      "id": 550771041
	}
  ]
}

```
My code looks like this
```
ValidationContext validationContext = new ValidationContext(uriFactory, urnFactory, JsonMetaSchema.getV7(), factory, config);
       RefValidator refValidator = new RefValidator("???", ???, ???, validationContext);
       refValidator.walk(inputData , iDontKnoWhatShouldBeHere , "????" , true);

```

No matter how I tried to do something, but it doesn’t work, there is no intelligible documentation, I’m sitting wondering what needs to be done to make it work.

**Comment by fdutton** [comment]
@vitno1 What is it that you are trying to do? The RefValidator is an internal component. Usually one uses JsonSchemaFactory to create a JsonSchema and then uses JsonSchema to either validate a JSON instance or walk the schema.

You can find more information about implementing a walker in this documentation. https://github.com/networknt/json-schema-validator/blob/master/doc/walkers.md

**Reply by vitno1** [reply]
Imagine that you have a scheme in which you have links to other schemes.
You are trying to transmit data in the form of json to do a check on several schemes.

I have read your documentation many times, but it is not informative, you have a lot of Public methods that do not even have a description in the source code.

**Reply by vitno1** [reply]
Sry for my bad english)

**Reply by vitno1** [reply]
Maybe there is at least some tutorial on using the RefValidator class, or some documentation?
The json-schema-validator documentation simply says, quote "The JSONValidator interface extends this new interface thus allowing all the validator's defined in library to implement this new interface. BaseJsonValidator class provides a default implementation of the walk method. In this case the walk method does nothing but validating based on shouldValidateSchema parameter."

Where are the examples of validator implementation, or a detailed description??

**Comment by fdutton** [comment]
I used the [Getting Started](https://github.com/networknt/json-schema-validator/blob/master/doc/quickstart.md) documentation when I first started using this library.

The problem that I see is you are attempting to use a component of the library that is meant for internal use. Try this instead
```
class MyClass {
    private final ObjectMapper mapper;
    private final JsonSchema schema;

    MyClass() {
        URI location = URI.create(/* location of first schema */);
        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);
        this.schema = factory.getSchema(location);
        this.mapper = new ObjectMapper();
    }

    void checkData(String data) {
        JsonNode json = this.mapper.readTree(data);
        Set<ValidationMessage> errors = this.schema.validate(json);
        // do something with the errors
    }
}
```

**Reply by vitno1** [reply]
You have json-schema which has links to other json-schemas and you need to use RefValidator

**Reply by vitno1** [reply]
Here are two links to another schema
{
  "$schema": "https://json-schema.org/draft-07/schema#",
  "title": "1234 some title",
  "type": "object",
  "properties": {
    "rowFirst": {
      "$ref": ".1234.json#/row"
    },
    "rowSecond": {
      "$ref": ".1234.json#/row"
    }
  },
  "required": [
    "rowFirst",
    "rowSecond"
  ]
}

Here's another schema

{
  "$schema": "https://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "row": {
      "type": "array",
      "title": "1234 someTitle",
      "items": {
        "properties": {
          "id": {
            "title": "identifier",
            "description": "some description",
            "type": "integer"
          },
          "year": {
            "title": "born year",
            "description": "",
            "type": "integer"
          }
        },
        "required": [
          "id",
          "year"
        ],
        "maxProperties": 2
      }
    }
  }
}

Incoming json to be validated

{
  "rowFirst": [
    {
      "year": 2023
	},
    {
      "year": 2023
	}
  ],
  "rowSecond": [
    { 
      "id": 550771041
	}
  ]
}

**Reply by vitno1** [reply]
RefValidator should solve this problem, but it doesn't work.
I have described the problem in sufficient detail, I do not know what may not be clear.

**Comment by fdutton** [comment]
`JsonSchema` uses `RefValidator` internally. It is not meant to be used separately, which is why I proposed that you try the example I provided.

You continue to say that, "it doesn't work" but you have not explained how it is not working.
